src/mac.cpp

```diff
@@ -0,0 +1,112 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2020 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#include "mococrw/mac.h"
+#include "mococrw/error.h"
+#include "mococrw/openssl_wrap.h"
+
+#include <openssl/hmac.h>
+
+namespace mococrw
+{
+
+class HMAC::Impl
+{
+public:
+    Impl(openssl::DigestTypes hashFunction, const std::vector<uint8_t> key, size_t outputLength) :
+        outputLength(outputLength)
+    {
+        const EVP_MD* digestFn = _getMDPtrFromDigestType(hashFunction);
+
+        if (key.empty()) {
+            throw MoCOCrWException("Key for HMAC is empty.");
+        }
+
+        ctx = openssl::_HMAC_CTX_new();
+        openssl::_HMAC_Init_ex(ctx.get(), key.data(), key.size(), digestFn, NULL);
+    }
+
+    ~Impl() = default;
+
+    void update(const std::vector<uint8_t> &message)
+    {
+        if (_isFinished) {
+            throw MoCOCrWException("update() can't be called after finish()");
+        }
+        openssl::_HMAC_Update(ctx.get(), message.data(), message.size());
+    }
+
+    std::vector<uint8_t> finish()
+    {
+        if (_isFinished) {
+            throw MoCOCrWException("finish() can't be called twice.");
+        }
+
+        /* Currently SHA512 has the greatest hash size */
+        _result.resize(EVP_MAX_MD_SIZE);
+
+        unsigned int tagLength;
+        openssl::_HMAC_Final(ctx.get(), _result.data(), &tagLength);
+
+        if (outputLength == 0) {
+            _result.resize(tagLength);
+        } else {
+            if (outputLength > tagLength) {
+                throw MoCOCrWException("output length shall be greater than the actual tag length.");
+            }
+            _result.resize(outputLength);
+        }
+
+        _isFinished = true;
+
+        return _result;
+    }
+
+private:
+    openssl::SSL_HMAC_CTX_SharedPtr ctx = nullptr;
+    std::vector<uint8_t> _result;
```

Comment:
I don't think `_result` needs to be a member here. You can create it in `finish()`.